java/ql/lib/semmle/code/java/security/SensitiveApi.qll

```diff
@@ -3,17 +3,45 @@
  */
 
 import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.ExternalFlow
+
+/**
+ * A node which represents the use of a credential.
+ */
+abstract class CredentialSinkNode extends DataFlow::Node { }
+
+/**
+ * A node representing a password being passed to a method.
+ */
+class PasswordSink extends CredentialSinkNode {
+  PasswordSink() { sinkNode(this, "credential-password") }
+}
+
+/**
+ * A node representing a username being passed to a method.
+ */
+class UsernameSink extends CredentialSinkNode {
+  UsernameSink() { sinkNode(this, "credential-username") }
+}
+
+/**
+ * A node representing a cryptographic key being passed to a method.
+ */
+class CryptoKeySink extends CredentialSinkNode {
+  CryptoKeySink() { sinkNode(this, "crypto-parameter") }
```

Comment:
I wasn't sure which way I think this kind should be named. The main issue here is that this type is currently doing double-duty for both

1. Cryptographic keys used for a form of authentication
2. Parameters to a cryptographic operation

So I defaulted to using the more generic "crypto-parameter" when I first wrote this.

Unless we think it's better to create two kinds:

| kind name                | use cases                                                                                   |
|--------------------------|---------------------------------------------------------------------------------------------|
| `credentials-crypto-key` | A value that can be used as a key for authentication                                        |
| `crypto-parameter`       | A value which is used for a cryptographic operation which is not necessarily authentication |

In which case I'll go through the models and distinguish between credential cryptographic parameters and other crypto params.

Reply:
Actually upon further inspection, since `encryption-%` is one of the things we support for kinds, it might make sense to label some of the non-credential crypto parameters accordingly.